content/documentation/tutorials/nats-docker-swarm.md

```diff
@@ -0,0 +1,95 @@
++++
+date = "2017-09-07"
+title = "Using NATS with Docker Swarm"
+description = ""
+category = "tutorials"
+[menu.main]
+  name = "Using NATS with Docker Swarm"
+  weight = 4
+  identifier = "tutorial-nats-docker-swarm"
+  parent = "Tutorials"
++++
+
+# Using NATS with Docker Swarm
+
+# Step 1:
+Create and overlay network for the cluster (in this example, `nats-cluster-example`), and instantiate an initial NATS server.
+
+First create an overlay network:
+```
+docker network create --driver overlay nats-cluster-example
+```
+
+Next instantiate an initial "seed" server for a NATS cluster:
+```
+docker service create --network nats-cluster-example --name nats-cluster-node-1 nats:1.0 -DV
+```
+
+# Step 2:
+The 2nd step is to create another service which connects to the NATS server within the overlay network.  Note that we have an initial IP for connecting to the server:
+
+```
+docker service create --name ruby-nats --network nats-cluster-example wallyqs/ruby-nats:ruby-2.3.1-nats-v1.0 -e '
```

Comment:
we could simplify this to not use ruby somehow but here image has to be `wallyqs/ruby-nats:ruby-2.3.1-nats-v0.8.0` :

```
docker service create --name ruby-nats --network nats-cluster-example wallyqs/ruby-nats:ruby-2.3.1-nats-v0.8.0 -e '
  NATS.on_error do |e|
    puts "ERROR: #{e}"
  end
  NATS.start(:servers => ["nats://nats-cluster-node-1:4222"]) do |nc|
    inbox = NATS.create_inbox
    puts "[#{Time.now}] Connected to NATS at #{nc.connected_server}, inbox: #{inbox}"

    nc.subscribe(inbox) do |msg, reply|
      puts "[#{Time.now}] Received reply - #{msg}"
    end

    nc.subscribe("hello") do |msg, reply|
      next if reply == inbox
      puts "[#{Time.now}] Received greeting - #{msg} - #{reply}"
      nc.publish(reply, "world")
    end

    EM.add_periodic_timer(1) do
      puts "[#{Time.now}] Saying hi (servers in pool: #{nc.server_pool}"
      nc.publish("hello", "hi", inbox)
    end
  end'
```